File: rts/clojure/src/rts/sensor.clj
(ns rts.sensor
  (:require [langohr.core :as rmq]
            [langohr.channel :as lch]
            [langohr.queue :as lq]
            [langohr.basic :as lb]
            [rts.utils :refer [state get-attr get-unit get-iso-dt ansi-green
                               ansi-reset rfill]]))

(defn run [name generator]
  (let [con (rmq/connect)
        ch (lch/open con)
        unit (get-unit name)]
    (lq/declare ch name {:exclusive false :auto-delete false})
    (loop []
      (when (> (:distance @state) 0)
        (let [data (generator)]
          (dosync (alter state assoc (get-attr name) data))
          (lb/publish ch "" name (str data))
          (printf "%s127.0.0.1 sensor %s [%s] \"measured value of %d%s\"%s\n"
                  ansi-green (rfill name) (get-iso-dt) data unit ansi-reset)
          (flush)
          (Thread/sleep 2000)
          (recur))))
    (rmq/close ch)
    (rmq/close con)
    (printf "127.0.0.1 sensor %s [%s] \"closing connection\"\n" name (get-iso-dt))
    (flush)))

File: rts/java/app/src/main/java/rts/Sensor.java
package rts;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.io.IOException;
import java.lang.InterruptedException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.TimeoutException;

public class Sensor implements Runnable {
  private ConnectionFactory factory;
  private State state;
  private String name;
  private DataGenerator generator;

  public Sensor(String name, State state, ConnectionFactory factory,
      DataGenerator generator) {
    this.name = name;
    this.state = state;
    this.factory = factory;
    this.generator = generator;
  }

  @Override
  public void run() {
    try (Connection con = factory.newConnection()) {
      Channel ch = con.createChannel();
      ch.queueDeclare(name, false, false, false, null);

      try {
        while (state.getDistance() > 0) {
          int data = generator.generate();
          state.set(name, data);
          ch.basicPublish("", name, null, String.valueOf(data).getBytes());
          System.out.printf(
              "%s127.0.0.1 sensor %s [%s] \"measured value of %d%s\"%s\n",
              Main.ANSI_GREEN, rfill(name), getISODateTime(), data, getUnit(),
              Main.ANSI_RESET);
          Thread.sleep(2000);
        }
      } catch (InterruptedException e) {
        System.err.println(e.getMessage());
      }
      System.out.printf("127.0.0.1 sensor %s [%s] \"closing connection\"\n",
          name, getISODateTime());
    } catch (TimeoutException | IOException e) {
      System.err.println(e.getMessage());
    }
  }

  private String getISODateTime() {
    LocalDateTime now = LocalDateTime.now();
    String dt = now.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
    return dt.substring(0, dt.lastIndexOf('.'));
  }

  private String rfill(String s) {
    return s + " ".repeat(8 - s.length());
  }

  private String getUnit() {
    switch (name) {
      case Main.QUEUE_GPS:
        return " km";
      case Main.QUEUE_ALTITUDE:
        return " m";
      case Main.QUEUE_SPEED:
        return " km/h";
      case Main.QUEUE_PRESSURE:
        return " pa";
      case Main.QUEUE_WEATHER:
        return "";
    }
    return "";
  }
}
